sources/sql_database/pg_replication/helpers.py

```diff
@@ -0,0 +1,528 @@
+        port=credentials.port,
+        connection_factory=connection_factory,
+    )
+
+
+def _get_rep_conn(
+    credentials: ConnectionStringCredentials,
+) -> LogicalReplicationConnection:
+    # returns a psycopg2 LogicalReplicationConnection
+    return _get_conn(credentials, LogicalReplicationConnection)  # type: ignore[return-value]
+
+
+def _make_qualified_table_name(table_name: str, schema_name: str) -> str:
+    return (
+        escape_postgres_identifier(schema_name)
+        + "."
+        + escape_postgres_identifier(table_name)
+    )
+
+
+def _gen_table_replication_references(
+    table_name: str, schema_name: str
+) -> Tuple[str, str]:
+    # generate replication slot and publication names dedicated to a single table
+    slot_name = f"_dlt_slot_{schema_name}_{table_name}"
+    publication_name = f"_dlt_pub_{schema_name}_{table_name}"
+    return (slot_name, publication_name)
+
+
+@dataclass
+class ItemGenerator:
+    credentials: ConnectionStringCredentials
+    slot_name: str
+    options: Dict[str, str]
+    upto_lsn: Optional[int] = None
+    include_columns: Optional[Sequence[str]] = None
+    last_commit_lsn: Optional[int] = field(default=None, init=False)
+
+    def __iter__(self) -> Iterator[Union[TDataItem, DataItemWithMeta]]:
+        """Consumes messages from replication slot and generates data items.
+
+        Does not advance the slot.
+        Maintains LSN of last consumed Commit message in object state.
+        """
+        cur = _get_rep_conn(self.credentials).cursor()
+        try:
+            consumed_all = False
+            consumer = MessageConsumer(self.upto_lsn, self.include_columns)
+            cur.start_replication(
+                slot_name=self.slot_name, decode=False, options=self.options
+            )
+            cur.consume_stream(consumer)
+        except StopReplication:
+            # all messages upto `upto_lsn` have been successfully consumed
+            consumed_all = True
+        finally:
+            cur.connection.close()
+            if consumed_all:
+                for i in consumer.data_items:
+                    yield i
+                self.last_commit_lsn = consumer.last_commit_lsn
+
+
+class MessageConsumer:
+    """Consumes messages from a ReplicationCursor.
+
+    Assumes all messages passed to __call__ have the same relation_id, i.e. they
+    belong to the same table.
+    """
+
+    def __init__(
+        self,
+        upto_lsn: int,
+        include_columns: Optional[Sequence[str]] = None,
+    ) -> None:
+        self.upto_lsn = upto_lsn
+        self.include_columns = include_columns
+
+        # data_items attribute maintains all data items
+        self.data_items: List[Union[TDataItem, DataItemWithMeta]] = []
+        # other attributes only maintain last-seen values
+        self.last_table_schema: TTableSchema
+        self.last_commit_ts: pendulum.DateTime
+        self.last_commit_lsn = None
+
+    def __call__(self, msg: ReplicationMessage) -> None:
```

Comment:
do I understand correctly that we get a message one by one? I hope that our python code will keep up. also we need some batching (as you say in the ticket)

Reply:
Yeah, messages come in one by one. I will implement batching and then do some load tests to see if performance is acceptable for larger volumes.